domain/annotation/state/state.go

```diff
@@ -0,0 +1,283 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/annotations"
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/domain"
+	statement "github.com/juju/juju/internal/database"
+)
+
+// State represents a type for interacting with the underlying state.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new State for interacting with the underlying state.
+func NewState(factory database.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// GetAnnotations will retrieve all the annotations associated with the given ID from the database.
+// If no annotations are found, an empty map is returned.
+func (st *State) GetAnnotations(ctx context.Context, id annotations.ID) (map[string]string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// Prepare query for getting the annotations of ID
+	getAnnotationsQuery := getAnnotationQueryForID(id)
+
+	getAnnotationsStmt, err := sqlair.Prepare(getAnnotationsQuery, Annotation{}, sqlair.M{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing get annotations query for ID: %q", id.Name)
+	}
+
+	if id.Kind == annotations.KindModel {
+		return st.getAnnotationsForModel(ctx, db, id, getAnnotationsStmt)
+	}
+	return st.getAnnotationsForID(ctx, db, id, getAnnotationsStmt)
+}
+
+// getAnnotationsForModel retrieves all annotations aassociated with the given model id from the
+// database. If no annotations are found, an empty map is returned. This is specialized as opposed
+// to the other Kinds because we keep annotations per model, so we don't need to try to find the
+// uuid of the given id (the model).
+func (st *State) getAnnotationsForModel(ctx context.Context, db database.TxnRunner, id annotations.ID, getAnnotationsStmt *sqlair.Statement) (map[string]string, error) {
+	// Running transactions for getting annotations
+	var annotationsResults []Annotation
+	err := db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return tx.Query(ctx, getAnnotationsStmt).GetAll(&annotationsResults)
+	})
+
+	if err != nil {
+		if errors.Is(err, sqlair.ErrNoRows) {
+			// No errors, we return empty map if no annotation is found
+			return map[string]string{}, nil
```

Comment:
Does an empty map mean something compared to nil?

Reply:
Not a big difference. I just kinda like the interface where we always return a `map[string]string{}` (i.e. object that's initialized), as opposed to return sometimes a `map[string]string{}` and sometimes a null pointer. That way for example the caller can safely run things like `len()` without having to check against a panic. I can be convinced either way. @SimonRichardson wdyt?

Reply:
> the caller can safely run things like len() without having to check against a panic

[Calling `len()` on a typed nil value doesn't panic](https://go.dev/play/p/qkA1KlBKeC_W).